src/decoder_opus.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * This file is part of EasyRPG Player.
+ *
+ * EasyRPG Player is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * EasyRPG Player is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "system.h"
+
+#ifdef HAVE_OPUS
+
+#include <opus/opusfile.h>
+#include "audio_decoder.h"
+#include "decoder_opus.h"
+#include "output.h"
+
+static int custom_read(void *stream, unsigned char *ptr, int nbytes) {
+	FILE* f = reinterpret_cast<FILE*>(stream);
+	return fread(ptr, 1, nbytes, f);
+}
+
+OpusDecoder::OpusDecoder() {
+	music_type = "opus";
+}
+
+OpusDecoder::~OpusDecoder() {
+	if (oof) {
+		op_free(oof);
+	}
+}
+
+bool OpusDecoder::Open(FILE* file) {
+	finished = false;
+
+	int res;
+	OpusFileCallbacks callbacks = {custom_read, nullptr, nullptr, nullptr};
+
+	oof = op_open_callbacks(file, &callbacks, nullptr, 0, &res);
+	if (res != 0) {
+		error_message = "Opus: Error reading file";
+		op_free(oof);
+		fclose(file);
+		return false;
+	}
+
+	return true;
+}
+
+bool OpusDecoder::Seek(size_t offset, Origin origin) {
+	if (offset == 0 && origin == Origin::Begin) {
+		if (oof) {
+			op_raw_seek(oof, 0);
+		}
+		finished = false;
+		return true;
+	}
+
+	return false;
+}
+
+bool OpusDecoder::IsFinished() const {
+	if (!oof)
+		return false;
+
+	return finished;
+}
+
+void OpusDecoder::GetFormat(int& freq, AudioDecoder::Format& format, int& chans) const {
+	freq = frequency;
+	format = Format::S16;
+	chans = channels;
+}
+
+bool OpusDecoder::SetFormat(int freq, AudioDecoder::Format format, int chans) {
+	if (freq != frequency || chans != channels || format != Format::S16)
+		return false;
+
+	return true;
+}
+
+int OpusDecoder::FillBuffer(uint8_t* buffer, int length) {
+	if (!oof)
+		return -1;
+
+	return op_read_stereo(oof, reinterpret_cast<opus_int16*>(buffer), length / 2) * 4;
```

Comment:
I have to think about the `/2` and `* 4` here.

---
The doc says:
`int op_read_stereo (OggOpusFile *_of, opus_int16 *_pcm, int _buf_size)`

Note: `_buf_size` indicates the total number of values that can be stored in `_pcm`, while the return value is the number of samples per channel, even though the channel count is known, for consistency with `op_read()`.

- `_pcm`: A buffer in which to store the output PCM samples, as signed native-endian 16-bit values at 48 kHz with a nominal range of [-32768,32767). The left and right channels are interleaved in the buffer. This must have room for at least `_buf_size` values.
- `_buf_size`: The number of values that can be stored in `_pcm`. It is recommended that this be large enough for at least 120 ms of data at 48 kHz per channel (11520 values total). Smaller buffers will simply return less data, possibly consuming more memory to buffer the data internally.
- Returns the number of samples read per channel on success, or a negative value on failure.
---

To me this means it will not fill more samples than `length/2` anyway, because it is stereo.
This means the buffer could be half-empty here.
The return value is the number of bytes read.
This should be the returned value (number of samples per channel) `*2` (for stereo) `*2` (because samples are 16 bit), so `*4` seems right here.
However, in case of error, the negative error value is multiplied as well. I do not know how the error constants are defined, but this could possibly lead to overflows. (`#define ERROR_CODE INT_MIN`)